File: frontend/src/types/clearRecord.js
/**
 * クリア記録関連の型定義（機体別条件対応）
 * JSDocを使用してTypeScriptライクな型チェックを提供
 */

/**
 * クリア記録
 * @typedef {Object} ClearRecord
 * @property {number} id - 記録ID
 * @property {number} user_id - ユーザーID
 * @property {number} game_id - ゲームID
 * @property {number} character_id - キャラクターID
 * @property {string} difficulty - 難易度
 * @property {boolean} is_cleared - 通常クリア
 * @property {boolean} is_no_continue_clear - ノーコンティニュークリア
 * @property {boolean} is_no_bomb_clear - ノーボムクリア
 * @property {boolean} is_no_miss_clear - ノーミスクリア
 * @property {boolean} is_full_spell_card - フルスペルカード取得
 * @property {boolean} is_special_clear_1 - 特殊クリア条件1（作品固有）
 * @property {boolean} is_special_clear_2 - 特殊クリア条件2（作品固有）
 * @property {boolean} is_special_clear_3 - 特殊クリア条件3（作品固有）
 * @property {string} [cleared_at] - クリア日（YYYY-MM-DD）
 * @property {string} [last_updated_at] - 最終更新日時（ISO文字列）
 * @property {string} [created_at] - 作成日時（ISO文字列）
 */

/**
 * クリア記録作成/更新データ
 * @typedef {Object} ClearRecordFormData
 * @property {number} game_id - ゲームID
 * @property {number} character_id - キャラクターID
 * @property {string} difficulty - 難易度
 * @property {boolean} is_cleared - 通常クリア
 * @property {boolean} is_no_continue_clear - ノーコンティニュークリア
 * @property {boolean} is_no_bomb_clear - ノーボムクリア
 * @property {boolean} is_no_miss_clear - ノーミスクリア
 * @property {boolean} is_full_spell_card - フルスペルカード取得
 * @property {boolean} is_special_clear_1 - 特殊クリア条件1（作品固有）
 * @property {boolean} is_special_clear_2 - 特殊クリア条件2（作品固有）
 * @property {boolean} is_special_clear_3 - 特殊クリア条件3（作品固有）
 * @property {string} [cleared_at] - クリア日（YYYY-MM-DD）
 */

/**
 * 機体別条件データ（タブ式UI用）
 * @typedef {Object} IndividualConditionData
 * @property {boolean} cleared - 通常クリア
 * @property {boolean} no_continue - ノーコンティニュークリア
 * @property {boolean} no_bomb - ノーボムクリア
 * @property {boolean} no_miss - ノーミスクリア
 * @property {boolean} full_spell_card - フルスペルカード取得
 * @property {boolean} special_clear_1 - 特殊クリア条件1（作品固有）
 * @property {boolean} special_clear_2 - 特殊クリア条件2（作品固有）
 * @property {boolean} special_clear_3 - 特殊クリア条件3（作品固有）
 */

/**
 * 難易度別データ（タブ式UI用）
 * @typedef {Object} DifficultyData
 * @property {Object.<string, IndividualConditionData>} characters - キャラクターID → 条件データのマップ
 */

/**
 * クリア条件の種類
 */
export const CLEAR_CONDITIONS = {
  CLEARED: 'cleared',
  NO_CONTINUE: 'no_continue',
  NO_BOMB: 'no_bomb',
  NO_MISS: 'no_miss',
  FULL_SPELL_CARD: 'full_spell_card',
  SPECIAL_CLEAR_1: 'special_clear_1',
  SPECIAL_CLEAR_2: 'special_clear_2',
  SPECIAL_CLEAR_3: 'special_clear_3'
};

/**
 * クリア条件の配列（map用）
 */
export const CLEAR_CONDITIONS_ARRAY = [
  CLEAR_CONDITIONS.CLEARED,
  CLEAR_CONDITIONS.NO_CONTINUE,
  CLEAR_CONDITIONS.NO_BOMB,
  CLEAR_CONDITIONS.NO_MISS,
  CLEAR_CONDITIONS.FULL_SPELL_CARD
];

/**
 * 対戦型STG用のクリア条件配列（フルスペカなし）
 */
export const CLEAR_CONDITIONS_ARRAY_VERSUS = [
  CLEAR_CONDITIONS.CLEARED,
  CLEAR_CONDITIONS.NO_CONTINUE,
  CLEAR_CONDITIONS.NO_BOMB,
  CLEAR_CONDITIONS.NO_MISS
];

/**
 * ゲームタイプに応じたクリア条件配列を取得
 * @param {string} gameType - ゲームタイプ
 * @param {number} gameId - ゲームID（特殊条件判定用）
 * @returns {string[]} クリア条件配列
 */
export const getClearConditionsForGameType = (gameType, gameId = null) => {
  let conditions = [];
  
  if (gameType === 'versus') {
    conditions = CLEAR_CONDITIONS_ARRAY_VERSUS;
  } else {
    conditions = CLEAR_CONDITIONS_ARRAY;
  }
  
  // 特殊クリア条件を持つゲーム
  if (gameId === 7) {
    // 星蓮船：ノーベントラー
    conditions = [...conditions, CLEAR_CONDITIONS.SPECIAL_CLEAR_1];
  } else if (gameId === 8) {
    // 妖精大戦争：ノーアイス
    conditions = [...conditions, CLEAR_CONDITIONS.SPECIAL_CLEAR_1];
  } else if (gameId === 9) {
    // 神霊廟：ノートランス
    conditions = [...conditions, CLEAR_CONDITIONS.SPECIAL_CLEAR_1];
  } else if (gameId === 12) {
    // 天空璋：ノー季節解放
    conditions = [...conditions, CLEAR_CONDITIONS.SPECIAL_CLEAR_1];
  } else if (gameId === 13) {
    // 鬼形獣：ノー暴走、ノー霊撃
    conditions = [...conditions, CLEAR_CONDITIONS.SPECIAL_CLEAR_1, CLEAR_CONDITIONS.SPECIAL_CLEAR_2];
  } else if (gameId === 14) {
    // 虹龍洞：ノーカード
    conditions = [...conditions, CLEAR_CONDITIONS.SPECIAL_CLEAR_1];
  }
  
  return conditions;
};

/**
 * クリア条件の表示ラベル
 */
export const CLEAR_CONDITION_LABELS = {
  [CLEAR_CONDITIONS.CLEARED]: 'クリア',
  [CLEAR_CONDITIONS.NO_CONTINUE]: 'ノーコンティニュー',
  [CLEAR_CONDITIONS.NO_BOMB]: 'ノーボム',
  [CLEAR_CONDITIONS.NO_MISS]: 'ノーミス',
  [CLEAR_CONDITIONS.FULL_SPELL_CARD]: 'フルスペカ',
  [CLEAR_CONDITIONS.SPECIAL_CLEAR_1]: '特殊条件1',
  [CLEAR_CONDITIONS.SPECIAL_CLEAR_2]: '特殊条件2',
  [CLEAR_CONDITIONS.SPECIAL_CLEAR_3]: '特殊条件3'
};

/**
 * ゲーム別の特殊クリア条件ラベル
 */
export const getSpecialClearLabel = (gameId, specialType) => {
  const labels = {
    7: { // 星蓮船
      special_clear_1: 'ノーベントラー'
    },
    8: { // 妖精大戦争
      special_clear_1: 'ノーアイス'
    },
    9: { // 神霊廟
      special_clear_1: 'ノートランス'
    },
    12: { // 天空璋
      special_clear_1: 'ノー季節解放'
    },
    13: { // 鬼形獣
      special_clear_1: 'ノー暴走',
      special_clear_2: 'ノー霊撃'
    },
    14: { // 虹龍洞
      special_clear_1: 'ノーカード'
    }
  };
  
  return labels[gameId]?.[specialType] || `特殊条件${specialType.slice(-1)}`;
};

/**
 * クリア条件の説明
 */
export const CLEAR_CONDITION_DESCRIPTIONS = {
  [CLEAR_CONDITIONS.CLEARED]: '通常クリア',
  [CLEAR_CONDITIONS.NO_CONTINUE]: 'コンティニューなしでクリア',
  [CLEAR_CONDITIONS.NO_BOMB]: 'ボムを使わずにクリア',
  [CLEAR_CONDITIONS.NO_MISS]: '被弾なしでクリア',
  [CLEAR_CONDITIONS.FULL_SPELL_CARD]: '全スペルカード取得',
  [CLEAR_CONDITIONS.SPECIAL_CLEAR_1]: '特殊クリア条件1',
  [CLEAR_CONDITIONS.SPECIAL_CLEAR_2]: '特殊クリア条件2',
  [CLEAR_CONDITIONS.SPECIAL_CLEAR_3]: '特殊クリア条件3'
};

/**
 * ゲーム別の特殊クリア条件説明
 */
export const getSpecialClearDescription = (gameId, specialType) => {
  const descriptions = {
    7: { // 星蓮船
      special_clear_1: 'ベントラーを使用せずにクリア'
    },
    8: { // 妖精大戦争
      special_clear_1: 'アイスバリアを使用せずにクリア'
    },
    9: { // 神霊廟
      special_clear_1: 'トランスを使用せずにクリア'
    },
    12: { // 天空璋
      special_clear_1: 'シーズンリリースを使用せずにクリア'
    },
    13: { // 鬼形獣
      special_clear_1: 'ロアリングモードを使用せずにクリア',
      special_clear_2: 'ハイパーモードを使用せずにクリア'
    },
    14: { // 虹龍洞
      special_clear_1: 'アビリティカードを使用せずにクリア'
    }
  };
  
  return descriptions[gameId]?.[specialType] || `特殊クリア条件${specialType.slice(-1)}`;
};

/**
 * クリア条件の色（Tailwind CSS）
 */
export const CLEAR_CONDITION_COLORS = {
  [CLEAR_CONDITIONS.CLEARED]: 'text-blue-600',
  [CLEAR_CONDITIONS.NO_CONTINUE]: 'text-green-600',
  [CLEAR_CONDITIONS.NO_BOMB]: 'text-orange-600',
  [CLEAR_CONDITIONS.NO_MISS]: 'text-red-600',
  [CLEAR_CONDITIONS.FULL_SPELL_CARD]: 'text-purple-600',
  [CLEAR_CONDITIONS.SPECIAL_CLEAR_1]: 'text-cyan-600',
  [CLEAR_CONDITIONS.SPECIAL_CLEAR_2]: 'text-pink-600',
  [CLEAR_CONDITIONS.SPECIAL_CLEAR_3]: 'text-indigo-600'
};

/**
 * クリア記録から達成済み条件を取得
 * @param {ClearRecord} clearRecord - クリア記録
 * @returns {string[]} 達成済み条件のキー配列
 */
export const getAchievedConditions = (clearRecord) => {
  const conditions = [];
  if (clearRecord.is_cleared) conditions.push(CLEAR_CONDITIONS.CLEARED);
  if (clearRecord.is_no_continue_clear) conditions.push(CLEAR_CONDITIONS.NO_CONTINUE);
  if (clearRecord.is_no_bomb_clear) conditions.push(CLEAR_CONDITIONS.NO_BOMB);
  if (clearRecord.is_no_miss_clear) conditions.push(CLEAR_CONDITIONS.NO_MISS);
  if (clearRecord.is_full_spell_card) conditions.push(CLEAR_CONDITIONS.FULL_SPELL_CARD);
  if (clearRecord.is_special_clear_1) conditions.push(CLEAR_CONDITIONS.SPECIAL_CLEAR_1);
  if (clearRecord.is_special_clear_2) conditions.push(CLEAR_CONDITIONS.SPECIAL_CLEAR_2);
  if (clearRecord.is_special_clear_3) conditions.push(CLEAR_CONDITIONS.SPECIAL_CLEAR_3);
  return conditions;
};

/**
 * クリア記録が何らかの条件を達成しているかチェック
 * @param {ClearRecord} clearRecord - クリア記録
 * @returns {boolean} 何らかの条件を達成している場合true
 */
export const hasAnyCondition = (clearRecord) => {
  return clearRecord.is_cleared || 
         clearRecord.is_no_continue_clear || 
         clearRecord.is_no_bomb_clear || 
         clearRecord.is_no_miss_clear ||
         clearRecord.is_full_spell_card ||
         clearRecord.is_special_clear_1 ||
         clearRecord.is_special_clear_2 ||
         clearRecord.is_special_clear_3;
};

File: backend/domain/entities/clear_record.py
"""
クリア記録エンティティ（機体別個別条件対応）
"""
from datetime import datetime, date
from typing import Optional


class ClearRecord:
    """機体別個別条件記録エンティティ"""
    
    def __init__(
        self,
        id: Optional[int] = None,
        user_id: int = 0,
        game_id: int = 0,
        character_name: str = "",           # 機体名
        difficulty: str = "",
        mode: str = "normal",               # ゲームモード（通常: normal、紺珠伝: legacy/pointdevice）
        is_cleared: bool = False,
        is_no_continue_clear: bool = False,
        is_no_bomb_clear: bool = False,
        is_no_miss_clear: bool = False,
        is_full_spell_card: bool = False,   # フルスペルカード取得（全作品共通）
        is_special_clear_1: bool = False,  # 作品特有条件1 (例: ノーロアリング)
        is_special_clear_2: bool = False,  # 作品特有条件2 (例: ノー季節解放)
        is_special_clear_3: bool = False,  # 作品特有条件3 (例: その他特殊条件)
        cleared_at: Optional[date] = None,
        last_updated_at: Optional[datetime] = None,
        created_at: Optional[datetime] = None
    ):
        self.id = id
        self.user_id = user_id
        self.game_id = game_id
        self.character_name = character_name
        self.difficulty = difficulty
        self.mode = mode
        self.is_cleared = is_cleared
        self.is_no_continue_clear = is_no_continue_clear
        self.is_no_bomb_clear = is_no_bomb_clear
        self.is_no_miss_clear = is_no_miss_clear
        self.is_full_spell_card = is_full_spell_card
        self.is_special_clear_1 = is_special_clear_1
        self.is_special_clear_2 = is_special_clear_2
        self.is_special_clear_3 = is_special_clear_3
        self.cleared_at = cleared_at
        self.last_updated_at = last_updated_at
        self.created_at = created_at
    
    def has_any_clear_condition(self) -> bool:
        """いずれかのクリア条件が達成されているか"""
        return (self.is_cleared or 
                self.is_no_continue_clear or 
                self.is_no_bomb_clear or 
                self.is_no_miss_clear or
                self.is_full_spell_card or
                self.is_special_clear_1 or
                self.is_special_clear_2 or
                self.is_special_clear_3)
    
    def get_achieved_conditions(self) -> list[str]:
        """達成済み条件のリストを取得"""
        conditions = []
        if self.is_cleared:
            conditions.append("cleared")
        if self.is_no_continue_clear:
            conditions.append("no_continue")
        if self.is_no_bomb_clear:
            conditions.append("no_bomb")
        if self.is_no_miss_clear:
            conditions.append("no_miss")
        if self.is_full_spell_card:
            conditions.append("full_spell_card")
        if self.is_special_clear_1:
            conditions.append("special_1")
        if self.is_special_clear_2:
            conditions.append("special_2")
        if self.is_special_clear_3:
            conditions.append("special_3")
        return conditions
    
    def get_achievement_count(self) -> int:
        """達成済み条件の数を取得"""
        return len(self.get_achieved_conditions())
    
    def to_dict(self) -> dict:
        """辞書形式に変換"""
        return {
            'id': self.id,
            'user_id': self.user_id,
            'game_id': self.game_id,
            'character_name': self.character_name,
            'difficulty': self.difficulty,
            'mode': self.mode,
            'is_cleared': self.is_cleared,
            'is_no_continue_clear': self.is_no_continue_clear,
            'is_no_bomb_clear': self.is_no_bomb_clear,
            'is_no_miss_clear': self.is_no_miss_clear,
            'is_full_spell_card': self.is_full_spell_card,
            'is_special_clear_1': self.is_special_clear_1,
            'is_special_clear_2': self.is_special_clear_2,
            'is_special_clear_3': self.is_special_clear_3,
            'cleared_at': self.cleared_at.isoformat() if self.cleared_at else None,
            'last_updated_at': self.last_updated_at.isoformat() if self.last_updated_at else None,
            'created_at': self.created_at.isoformat() if self.created_at else None
        }
    
    @classmethod
    def from_dict(cls, data: dict) -> 'ClearRecord':
        """辞書からエンティティを作成"""
        return cls(
            id=data.get('id'),
            user_id=data.get('user_id', 0),
            game_id=data.get('game_id', 0),
            character_name=data.get('character_name', ''),
            difficulty=data.get('difficulty', ''),
            mode=data.get('mode', 'normal'),
            is_cleared=data.get('is_cleared', False),
            is_no_continue_clear=data.get('is_no_continue_clear', False),
            is_no_bomb_clear=data.get('is_no_bomb_clear', False),
            is_no_miss_clear=data.get('is_no_miss_clear', False),
            is_full_spell_card=data.get('is_full_spell_card', False),
            is_special_clear_1=data.get('is_special_clear_1', False),
            is_special_clear_2=data.get('is_special_clear_2', False),
            is_special_clear_3=data.get('is_special_clear_3', False),
            cleared_at=data.get('cleared_at'),
            last_updated_at=data.get('last_updated_at'),
            created_at=data.get('created_at')
        )
    
    def __str__(self) -> str:
        conditions = self.get_achieved_conditions()
        condition_str = ", ".join(conditions) if conditions else "未クリア"
        return f"ClearRecord(user={self.user_id}, game={self.game_id}, char={self.character_name}, diff={self.difficulty}, mode={self.mode}, conditions={condition_str})"
    
    def __repr__(self) -> str:
        return self.__str__()